net/speedtest/speedtest_server.go

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package speedtest
+
+import (
+	"crypto/rand"
+	"encoding/json"
+	"fmt"
+	"net"
+	"time"
+)
+
+// Serve starts up the server on a given host and port pair. It starts to listen for
+// connections and handles each one in a goroutine. Because it runs in an infinite loop,
+// this function only returns if any of the speedtests return with errors, or if a bool is sent
+// to the killSignal channel.
+func Serve(l *net.TCPListener, maxConnections int, killSignal chan bool, resultsChan chan []Result) error {
```

Comment:
"killSignal" is traditionally named "done". Or better yet this could accept a context.Context instead.